src/Symfony/Component/Form/FormFactory.php

```diff
@@ -62,48 +62,29 @@ public function createForProperty($class, $property, $data = null, array $option
     public function createBuilder($type = 'Symfony\Component\Form\Extension\Core\Type\FormType', $data = null, array $options = array())
     {
         $name = null;
-        $typeName = null;
 
         if ($type instanceof ResolvedFormTypeInterface) {
-            if (method_exists($type, 'getBlockPrefix')) {
-                // As of Symfony 3.0, the block prefix of the type is used as
-                // default name
-                $name = $type->getBlockPrefix();
-            } else {
-                // BC
-                $typeName = $type->getName();
-            }
+            $typeObject = $type;
         } elseif ($type instanceof FormTypeInterface) {
-            if (method_exists($type, 'getBlockPrefix')) {
-                // As of Symfony 3.0, the block prefix of the type is used as
-                // default name
-                $name = $type->getBlockPrefix();
-            } else {
-                // BC
-                $typeName = $type->getName();
-            }
+            $typeObject = $type;
         } elseif (is_string($type)) {
-            $resolvedType = $this->registry->getType($type);
-            if (method_exists($resolvedType, 'getBlockPrefix')) {
-                // As of Symfony 3.0, the block prefix of the type is used as
-                // default name
-                $name = $resolvedType->getBlockPrefix();
-            } else {
-                // BC
-                $typeName = $type;
-            }
+            $typeObject = $this->registry->getType($type);
+            $name = $type;
```

Comment:
I don't think you need to store the name here (i.e. the FQCN), as you want to get the block prefix from the type, no?